lib/bolt/pal.rb

```diff
@@ -286,15 +286,26 @@ def parse_manifest(code, filename)
       raise Bolt::PAL::PALError, "Failed to parse manifest: #{e}"
     end
 
-    def list_tasks
+    # Filters content by a list of names and glob patterns specified in project
+    # configuration.
+    def filter_content(content, patterns)
+      return content unless content && patterns
+
+      content.select do |name,|
+        patterns.any? { |pattern| File.fnmatch?(pattern, name, File::FNM_EXTGLOB) }
+      end
+    end
+
+    def list_tasks(filter_content: false)
       in_bolt_compiler do |compiler|
-        tasks = compiler.list_tasks
-        tasks.map(&:name).sort.each_with_object([]) do |task_name, data|
+        tasks = compiler.list_tasks.map(&:name).sort.each_with_object([]) do |task_name, data|
           task_sig = compiler.task_signature(task_name)
           unless task_sig.task_hash['metadata']['private']
             data << [task_name, task_sig.task_hash['metadata']['description']]
           end
         end
+
+        filter_content ? filter_content(tasks, @project&.tasks) : tasks
```

Comment:
I think any _other_ kind of filtering ought to just be done by the caller. The caller could always use the `filter_content` method directly to get that behavior if they want it. But generic filtering seems out of the scope of _this_ method, otherwise all it's doing is taking a value to pass through to another method, which seems like an unnecessary inversion of control.

Reply:
So, sorry, just to make sure I follow: PE needs to use this to filter tasks as well, but PE doesn't have projects right? I'm confused about this line: "And bolt server would be able to take advantage of that argument as well, since we're already separately planning on making it not return content that isn't listed in the project (rather than marking the content as allowed or not).". Are y'all using the same config settings as a blocklist instead of allowlist?

Reply:
PE does have projects. Currently, requests to bolt-server to list tasks\* in a project include _all_ tasks and include a field `"allowed": true/false` for each item to indicate whether the task is on the list in `bolt-project.yaml`. We want to change it to _only_ return the things on the list, rather than listing everything with an `allowed` field. If that filtering logic is in PAL, then bolt-server can just  call `pal.list_tasks(filter_content: true)` rather than having to do the filter itself. So if we make a change to how the filtering works, as this PR does, bolt-server will automatically be able to honor those rules.

\* This all applies to plans as well but I didn't want to repeat tasks/plans everywhere

Reply:
Ok that makes sense. I had thought y'all were adding a new way of filtering and didn't load project, which was why passing in the filter regex seemed useful.

This seems to do the right thing if `@project` is nil too, so if there are tools that try to use this without a project it should behave correctly.